compiler/rustc_typeck/src/astconv/mod.rs

```diff
@@ -1932,12 +1932,24 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
                 self.prohibit_generics(path.segments);
                 tcx.types.self_param
             }
-            Res::SelfTy(_, Some(def_id)) => {
+            Res::SelfTy(_, Some((def_id, forbid_generic))) => {
                 // `Self` in impl (we know the concrete type).
                 assert_eq!(opt_self_ty, None);
                 self.prohibit_generics(path.segments);
                 // Try to evaluate any array length constants.
-                self.normalize_ty(span, tcx.at(span).type_of(def_id))
+                let normalized_ty = self.normalize_ty(span, tcx.at(span).type_of(def_id));
+                if forbid_generic && normalized_ty.needs_subst() {
+                    tcx.sess
+                        .struct_span_err(
+                            path.span,
+                            "generic `Self` types are currently not permitted in anonymous constants"
+                        )
+                        .span_note(tcx.def_span(def_id), "not a concrete type")
```

Comment:
How can I get the `Span` of just the self type here? This currently returns the whole impl block

Reply:
This is the only issue with the PR that I see, but I can't think of the right way to get the span… cc @estebank

Reply:
```rust
if let Some(Node::Item(&hir::Item { kind: hir::ItemKind::Impl { self_ty, .. }, .. })) = tcx.hir().get_if_local(def_id) {
    err.span_note(self_ty.span, "not a concrete type");
}
```

Reply:
:+1:  thanks